src/psykal/psy.f90

```diff
@@ -0,0 +1,265 @@
+    real*8, intent(in)  :: pun(:,:,:), pvn(:,:,:), mydomain(:,:,:), zind(:,:,:), zslpx(:,:,:), zslpy(:,:,:)
+    real*8, intent(in)  :: zdt
+    integer, intent(in) :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji   
+    !
+    DO jk = 1, jpk-1
+       DO jj = 2, jpj-1
+          DO ji = 2, jpi-1
+             call zwxy2_kern(zwx,zwy,zdt,pun,pvn,mydomain,zind,zslpx,zslpy,jk,jj,ji)
+          END DO
+       END DO
+    END DO
+    !
+  end subroutine zwxy2_psy
+  !
+  subroutine mydomain_update_psy(mydomain,zwx,zwy,jpk,jpj,jpi)
+    !
+    use mydomain_update_kern_mod, only : mydomain_update_kern
+    !
+    real*8, intent(inout) :: mydomain(:,:,:)
+    real*8, intent(in)    :: zwx(:,:,:), zwy(:,:,:)
+    integer, intent(in)   :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji
+    !
+    DO jk = 1, jpk-1
+       DO jj = 2, jpj-1     
+          DO ji = 2, jpi-1
+             call mydomain_update_kern(mydomain,zwx,zwy,jk,jj,ji)
+          END DO
+       END DO
+    END DO
+    !
+  end subroutine mydomain_update_psy
+  !
+  subroutine zwx_psy(zwx,tmask,mydomain,jpk,jpj,jpi)
+    !
+    use zwx_kern_mod, only : zwx_kern
+    !
+    real*8, intent(out) :: zwx(:,:,:)
+    real*8, intent(in) :: tmask(:,:,:), mydomain(:,:,:)
+    integer, intent(in) :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji
+    !
+    DO jk = 2, jpk-1
+       DO jj = 1, jpj
+          DO ji = 1, jpi
+             call zwx_kern(zwx,tmask,mydomain,jk,jj,ji)
+          END DO
+       END DO
+    END DO
+    !
+  end subroutine zwx_psy
+  !
+  subroutine zslpx_psy(zslpx,zwx,jpk,jpj,jpi)
+    !
+    use zslpx_kern_mod, only : zslpx_kern
+    !
+    real*8, intent(out) :: zslpx(:,:,:)
+    real*8, intent(in) :: zwx(:,:,:)
+    integer, intent(in) :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji
+    !
+    DO jk = 2, jpk-1
+       DO jj = 1, jpj
+          DO ji = 1, jpi
+             call zslpx_kern(zslpx,zwx,jk,jj,ji)
+          END DO
+       END DO
+    END DO
+    !
+  end subroutine zslpx_psy
+  !
+  subroutine zslpx_update_psy(zslpx,zwx,jpk,jpj,jpi)
+    !
+    use zslpx_update_kern_mod, only : zslpx_update_kern
+    !
+    real*8, intent(inout) :: zslpx(:,:,:)
+    real*8, intent(in) :: zwx(:,:,:)
+    integer, intent(in) :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji
+    !
+    DO jk = 2, jpk-1     
+       DO jj = 1, jpj
+          DO ji = 1, jpi
+             call zslpx_update_kern(zslpx,zwx,jk,jj,ji)
+          END DO
+       END DO
+    END DO
+    !
+  end subroutine zslpx_update_psy
+  !
+  subroutine zwx2_psy(zwx,zdt,zbtr,pwn,mydomain,zind,zslpx,jpk,jpj,jpi)
+    !
+    use zwx2_kern_mod, only : zwx2_kern
+    !
+    real*8, intent(inout) :: zwx(:,:,:)
+    real*8, intent(in) :: pwn(:,:,:), mydomain(:,:,:), zind(:,:,:), zslpx(:,:,:)
+    real*8, intent(in) :: zdt,zbtr
+    integer, intent(in) :: jpk,jpj,jpi
+    ! local variables
+    integer :: jk,jj,ji
+    !
+    DO jk = 1, jpk-1
```

Comment:
Loop bounds match those in original code. However that then means that the kernel updates points on (ji,jj,jk+1).